### DISCUSSION TITLE
Change how integer division works

### DISCUSSION THREAD

**Discussion by miv391** [godotengine/godot-proposals#6160] [Scripting]
A recurring question in Godot's Reddit channel is that why `1 / 2` (or something similar) returns `0`. Integer division is not a common concept in all languages, especially in script languages. I guess GDScript has integer division only because of C++.

My proposal is that the division operator `/` should always return a float, no matter what the operand types are (floats or integers). `1 / 2` would return float `0.5`. `4 / 2` would return float `2.0`. 

A new integer division operator should be added. It could be for example `//`. `1 // 2` woud return `0`, `4 // 2` would return `2`. Using integer division with any other operand types than two integers would be an error.

Now using the integer division would always be a decision the programmer does knowingly, not by accident.

**Comment by AThousandShips** [comment]
Why is this an improvement over just `1/2.0`

This is the way it works in both Java and C# btw, it's actually common in programming languages from what I can find, where have you seen different?

**Reply by miv391** [reply]
A lot of people who use GDScript do not know that they must write `.0` to get simple math to work. I have programmed with C, so when I make the division 1/2 and get result 0, I know why and I can fix my error. I don't expect every Godot user having experience with C. If I make the same division with large numbers, say 37572245/7 and get result 5367463, I most probably do not notice that the result is not exactly what my intention was.

Sometimes integer divison is needed, but so seldom that in my opinion it shouldn't be the default.

At least in Javascript, Typescript, Python and Lua 1/2 results 0.5. In my opinion those are quite common languages.

**Comment by dalexeev** [comment]
* See #1866.

**Comment by TheSpartanK** [comment]
I've just come across this because of a bug, only to find 7/3 = 2 !!! 
Integer division is the most ridiculous thing I've eve come across. I'm not saying it's not valid in some obscure cases but to be the default is just insane.
I thought godot was good until I came across this madness, who decided this was a good idea?

**Comment by TheSpartanK** [comment]
This has been marked as closed but I don't see any resolution. Godot and GDScript in general are both very good but this issue is the worst aspect of it. Having integer division as the default is causing so many errors in my code, even though I'm aware of this stupid practice.

Can someone explain how it makes any sense to have integer division as the default, why would anyone want (by default) 50/100=0 or 1/3 = 0. Who thought this was a good idea in the first place and how is it still like this?

**Reply by miv391** [reply]
I closed this because to me it seems quite clear that the integer division handling is not going be changed. But I don't mind reopening this.

**Reply by TheSpartanK** [reply]
As a standard user of Godot, this is the most serious issue that I have encountered and I feel like it should not be overlooked. I know there are other threads regarding this issue but I think it's serious enough to keep pushing it, even on multiple threads. I understand the issue with back compatibility, changing this in Godot4 but waiting until Godot5 is not an acceptable solution.

I have 2 work arounds:

1) Keep integer division using '/' and add a normal division using a different symbol, i.e. '//' if it isn't already used
2) Give us a mod that we can manually install that either:
a) Gives the functionality in 1) or;
b) Changes the use of '/' to be normal division, not integer division

**Reply by sancho2** [reply]
I just want to note that FreeBasic used the forward slash / for standard(float) division and the backslash \ for integer division.   
(It also has operator overloading but that is neither here nor there)

**Reply by Joy-less** [reply]
@TheSpartanK If you are having such big problems with this, maybe you could try changing the way you divide numbers? Instead of using `a / b`, always use `a as float / b`?
```gdscript
print(3 as float / 4) # 0.75
```
@sancho2 Backslash should never be used outside of escape sequences in my opinion. Windows already made the mistake of using backslashes in file paths and it always causes problems when they're inserted into string literals (`"a\b\c"`).

**Comment by Joy-less** [comment]
There is a technical philosophical reason to not use `//` for integer division even though it's more convenient.

Normally, every arithmetic operation only involves a single number type. For example, `int + int = int` and `float * float = float`. In languages with multiple number types, this makes sense (`float64 / float64 = float64`).

However, adding an integer division operator means that number types need to be "aware" of an integer type. In other words, `float // float = int`. Integer division would be the only operation that returns a number of a different type. Of course you could implement it as `float // float = truncated float` but that would not resolve the `int / int = int` footgun.

**Reply by Mickeon** [reply]
`>>` and `<<` are also arithmetic operations that are "aware" of an integer type, which divide and multiply by 2, respectively.

**Reply by Joy-less** [reply]
> `>>` and `<<` are also arithmetic operations that are "aware" of an integer type, which divide and multiply by 2, respectively.

<< and >> do not operate on floats.
```gdscript
print(1.5 << 2.7) # Invalid operands to operator <<, float and float.
```
